File: protocols/uim-handshake/v1.0/examples/python/uim_client.py
"""
UIM Protocol Reference Implementation - Python
"""

import requests
import json
import time
import uuid

class UIMClient:
    def __init__(self, node_id, base_url, capabilities, solar_endpoint=None):
        self.node_id = node_id
        self.base_url = base_url
        self.capabilities = capabilities
        self.solar_endpoint = solar_endpoint
        self.connected_nodes = {}
        
    def generate_hello_response(self):
        """Generate a HELLO_RESPONSE for this node"""
        return {
            "node_id": self.node_id,
            "api_endpoint": f"{self.base_url}/uim",
            "capabilities": self.capabilities,
            "protocol_version": "UIM-HS-1.0",
            "public_key": "-----BEGIN PUBLIC KEY-----\n...\n-----END PUBLIC KEY-----",  # Placeholder
            "solar_endpoint": self.solar_endpoint
        }
    
    def discover_nodes(self, registry_url, capability_filter=None):
        """Discover other UIM nodes from registry"""
        try:
            response = requests.get(registry_url, timeout=10)
            nodes = response.json()
            
            if capability_filter:
                nodes = [node for node in nodes 
                        if capability_filter in node.get('capabilities', [])]
            
            self.connected_nodes = {node['node_id']: node for node in nodes}
            return nodes
            
        except requests.RequestException as e:
            print(f"Discovery failed: {e}")
            return []
    
    def send_task_proposal(self, target_node, task_type, input_context, max_solar_budget=0.001):
        """Send a task proposal to another UIM node"""
        task_proposal = {
            "task_id": f"task_{uuid.uuid4().hex[:8]}",
            "proposing_node": self.node_id,
            "task_type": task_type,
            "input_context": input_context,
            "expected_output_format": "markdown_bullets",
            "max_solar_budget": max_solar_budget,
            "required_rights_alignment": ["privacy", "non_discrimination"]
        }
        
        try:
            target_endpoint = f"{target_node['api_endpoint']}/task"
            response = requests.post(target_endpoint, 
                                   json=task_proposal,
                                   timeout=30)
            return response.json()
        except requests.RequestException as e:
            return {"error": str(e), "task_id": task_proposal["task_id"]}

# Example usage
if __name__ == "__main__":
    client = UIMClient(
        node_id="test-node-python-001",
        base_url="https://api.example.com",
        capabilities=["text-generation", "reasoning"],
        solar_endpoint="https://api.example.com/solar"
    )
    
    print("Hello Response:", json.dumps(client.generate_hello_response(), indent=2))

File: public/protocols/uim-handshake/v1.0/examples/javascript/uim-client.js
/**
 * UIM Protocol Reference Implementation - JavaScript
 */

class UIMClient {
    constructor(nodeId, baseUrl, capabilities, solarEndpoint = null) {
        this.nodeId = nodeId;
        this.baseUrl = baseUrl;
        this.capabilities = capabilities;
        this.solarEndpoint = solarEndpoint;
        this.connectedNodes = new Map();
    }

    generateHelloResponse() {
        return {
            node_id: this.nodeId,
            api_endpoint: `${this.baseUrl}/uim`,
            capabilities: this.capabilities,
            protocol_version: "UIM-HS-1.0",
            public_key: "-----BEGIN PUBLIC KEY-----\\n...\\n-----END PUBLIC KEY-----",
            solar_endpoint: this.solarEndpoint
        };
    }

    async discoverNodes(registryUrl, capabilityFilter = null) {
        try {
            const response = await fetch(registryUrl);
            const nodes = await response.json();
            
            const filteredNodes = capabilityFilter 
                ? nodes.filter(node => node.capabilities?.includes(capabilityFilter))
                : nodes;
            
            filteredNodes.forEach(node => {
                this.connectedNodes.set(node.node_id, node);
            });
            
            return filteredNodes;
        } catch (error) {
            console.error('Discovery failed:', error);
            return [];
        }
    }

    async sendTaskProposal(targetNode, taskType, inputContext, maxSolarBudget = 0.001) {
        const taskProposal = {
            task_id: `task_${Math.random().toString(36).substr(2, 9)}`,
            proposing_node: this.nodeId,
            task_type: taskType,
            input_context: inputContext,
            expected_output_format: "markdown_bullets",
            max_solar_budget: maxSolarBudget,
            required_rights_alignment: ["privacy", "non_discrimination"]
        };

        try {
            const response = await fetch(`${targetNode.api_endpoint}/task`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(taskProposal)
            });
            
            return await response.json();
        } catch (error) {
            return { error: error.message, task_id: taskProposal.task_id };
        }
    }
}

// Example usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = UIMClient;
    
    // Demo
    const client = new UIMClient(
        "test-node-js-001",
        "https://api.example.com",
        ["text-generation", "code-execution"],
        "https://api.example.com/solar"
    );
    
    console.log("Hello Response:", JSON.stringify(client.generateHelloResponse(), null, 2));
}
